### ISSUE TITLE
[ENH] design discussion - parameter validation and initialisation

### ISSUE THREAD

**Issue by fkiraly** [alan-turing-institute/sktime#1798]
The orthodox `sklearn` faith dictates that nothing must be done in `__init__` except writing parameters to `self`.
This does prevent side effects and careless overwrites, but I think this may be too harsh.

There are some negative consequences to this design:
* the user is only told that parameter settings are invalid when they try to `fit`. This is an unnatural point in the workflow, the natural point would be as soon as they attempt to construct the estimator with invalid parameters.
* `fit` needs to take the responsibility of constructing any component estimators, which seems unnatural. The natural nesting logic (for a simple composite) has `fit` call component `fit`-s, `predict` call component `predict`-s - and constructor call component constructors.
* for objects that do not have `fit`, such as metrics and distances, it is very unclear where input validation should even sit. There may be multiple methods that can be the first to call, so should the same boilerplate code sit in every function that could be called after construction? Seems unnatural.

What I would propose, in intentional deviation from `sklearn` doctrine:
* Two new interface points are added, abstract methods `validate_parameters` and `construct_components`. Neither must overwrite `__init__` parameter attributes, and both are called by `BaseObject`'s `__init__`. Individual estimator will implement the two, default behaviour is thet empty function.
* `validate_parameters` must have no side effects on the object scope whatsoever. Its purpose is exclusively to validate the parameters, and raise an exceptions if they are invalid. (it does not call estimator components' `validate_parameters`, since these have been called when the component objects have been constructed)
* `construct_components` creates any clones of estimator components used inside the object.

**Comment by fkiraly** [comment]
Any thoughts on this, @aiwalter and @TonyBagnall?

Also, @lmmentel and I had a discussion about this recently, in the context of `FeatureUnion`. @lmmentel also mentioned this in #1771. It came up in the metrics/distances as well, tagging @chrisholder and @eenticott-shell for that.

**Comment by Lovkush-A** [comment]
My initial thought is that this is a good idea. Note, I do not have a good sense of 'long-term' consequences or 'second-order' effects of such a change.

**Comment by fkiraly** [comment]
> My initial thought is that this is a good idea. Note, I do not have a good sense of 'long-term' consequences or 'second-order' effects of such a change.

I guess the main negatives are:
* more methods in the extender contract (user contract stays the same). One could now argue whether the "more methods" in the extender contract make it easier to understand, or more complex to comply with. I'd be on the side that, in the balance of things, it makes it easire and more intuitive to extend, but I see why one could be of the opposite opinion.
* we're doing it differently from `sklearn`, effectively committing heresy. Extenders may be confused only because we are different from `sklearn` which has a very heavy footprint.

Not sure whether there are other negatives? Would encourage to shoot at the proposal.

**Comment by TonyBagnall** [comment]
to me, the primary reason for conforming to scikit design is so that our estimators can be used in their utilities such as cross validation. If this change means that is not possible, I am dead against it. And you may think it wont be a factor, but there are certain idiosyncrasies with scikit (Im no expert, but certain things dont work as I woul have expected, such as distances in knn), so I would like it tested thoroughly before inclusion. I am less concerned with the look of conformity, since as I understand it (limited capacity to think it htrough atm) most of the changes are internal and additional, so the user can ignore if they want.

**Comment by fkiraly** [comment]
> I am less concerned with the look of conformity, since as I understand it (limited capacity to think it htrough atm) most of the changes are internal and additional, so the user can ignore if they want.

Yes, that's also what I would expect towards `scikit-learn`. I.e., I wouldn't expect any interoperability issues, but we'd have to see.
